nanoFramework.Device.Bluetooth/IO/DataReader.cs

```diff
@@ -0,0 +1,296 @@
+        public Guid ReadGuid()
+        {
+            byte[] byteArray = new byte[16];
+
+            // read position update and check are performed on the call
+            ReadBytes(byteArray);
+
+            return new Guid(byteArray);
+        }
+
+        /// <summary>
+        /// Reads a 16-bit integer value from the input buffer.
+        /// </summary>
+        /// <returns>The value.</returns>
+        public short ReadInt16()
+        {
+            var value = BitConverter.ToInt16(_buffer.Data, IncreaseReadPosition(2));
+
+            CheckReadPosition();
+
+            return value;
+        }
+
+        /// <summary>
+        /// Reads a 32-bit integer value from the input buffer.
+        /// </summary>
+        /// <returns>The value.</returns>
+        public int ReadInt32()
+        {
+            var value = BitConverter.ToInt32(_buffer.Data, IncreaseReadPosition(4));
+
+            CheckReadPosition();
+
+            return value;
+        }
+
+        /// <summary>
+        /// Reads a 64-bit integer value from the input buffer.
+        /// </summary>
+        /// <returns>The value.</returns>
+        public long ReadInt64()
+        {
+            var value = BitConverter.ToInt64(_buffer.Data, IncreaseReadPosition(8));
+
+            CheckReadPosition();
+
+            return value;
+        }
+
+        /// <summary>
+        /// Reads a floating-point value from the input buffer.
+        /// </summary>
+        /// <returns>The value.</returns>
+        public float ReadSingle()
+        {
+            var value = BitConverter.ToSingle(_buffer.Data, IncreaseReadPosition(4));
+
+            CheckReadPosition();
+
+            return value;
+        }
+
+        /// <summary>
+        /// Reads a string value from the input buffer.
+        /// </summary>
+        /// <param name="codeUnitCount">The length of the string.</param>
+        /// <returns>The value.</returns>
+        public string ReadString(UInt32 codeUnitCount)
+        {
+            Char[] buffer = new Char[codeUnitCount];
+
+            int readPosition = IncreaseReadPosition((int)codeUnitCount);
+
+            Encoding.UTF8.GetDecoder().Convert(_buffer.Data, readPosition, (int)codeUnitCount, buffer, 0, (int)codeUnitCount, false, out Int32 bytesUsed, out Int32 charsUsed, out Boolean completed);
+            var value = new String(buffer, 0, charsUsed);
+
+            CheckReadPosition();
+
+            return value;
+        }
+
+        /// <summary>
+        /// Reads a time interval from the input buffer.
+        /// </summary>
+        /// <returns>The value.</returns>
+        public TimeSpan ReadTimeSpan()
+        {
+            // read position update and check are performed on the call
+            return new TimeSpan(ReadInt64());
+        }
+
+        /// <summary>
+        /// Reads a 16-bit unsigned integer from the input buffer.
+        /// </summary>
+        /// <returns>The value.</returns>
+        public ushort ReadUInt16()
```

Comment:
note: RzadUInt16 is perfectly valid and great way. The remarks on UInt16 and others is just for variables. For function names, that's the right pattern and all perfect.

Reply:
Understood